drivers: STM32 async flash erase and write

jpowen898 opened the pull request:
- Implemented async flash erase and writes using the STM32 hal `HAL_FLASHEx_Erase_IT()` and `HAL_FLASH_Program_IT()` functions with their associated interrupt handler.
- I only implemented this for the h7, l4, and f4 stm32 families because that is the hardware I had available to test on.
- Async flash transactions can be enabled using the `FLASH_STM32_ASYNC` config option.
- Async transactions are implemented using a semaphore which blocks while waiting for the flash IRQ to indicate the end of a flash transaction. This allows other threads to take action while flash erase and write actions are taking place.

nordicjm commented on drivers/flash/Kconfig:
```diff
@@ -17,6 +17,11 @@ config FLASH_HAS_EX_OP
+config FLASH_HAS_ASYNC_OPS
+	bool
+	help
+		This option is selected by drivers that support asynchronous flash operations.
```
1x tab followed by 2x spaces for help text indent

nordicjm commented on drivers/flash/Kconfig.stm32:
```diff
@@ -25,6 +25,9 @@ menuconfig SOC_FLASH_STM32
 	select USE_STM32_HAL_FLASH if BT_STM32WBA
 	select USE_STM32_HAL_FLASH_EX if BT_STM32WBA
+	select FLASH_HAS_ASYNC_OPS if SOC_SERIES_STM32L4X || \
+								  SOC_SERIES_STM32F4X || \
+								  SOC_SERIES_STM32H7X
```
indent here is off

jpowen898 commented on drivers/flash/Kconfig.stm32:
```diff
@@ -100,4 +103,16 @@ config USE_MICROCHIP_QSPI_FLASH_WITH_STM32
+	select FLASH_HAS_ASYNC_OPS
+	select USE_STM32_HAL_FLASH
+	select USE_STM32_HAL_FLASH_EX
+	default y
```
Sounds good. I changed it to `default n`, and also added the semaphore to the `flash_stm32_read()` function to prevent the race conditions.

nordicjm replied:
Since flash in zephyr is expected to be blocking then I assume this will break a lot of things...

nordicjm commented on drivers/flash/Kconfig.stm32:
```diff
@@ -100,4 +103,14 @@ config USE_MICROCHIP_QSPI_FLASH_WITH_STM32
+	depends on MULTITHREADING && FLASH_HAS_ASYNC_OPS
+	select USE_STM32_HAL_FLASH
+	select USE_STM32_HAL_FLASH_EX
+	default n
```
```suggestion
```

erwango reviewed (changes requested):
As a general comment, please follow existing code style in this driver (and zephyr code base generally): break a line between code blocks.

Then it would be interesting to introduce this new feature into `doc/releases/release-notes-4.3.rst`

jpowen898 commented on drivers/flash/flash_stm32l4x.c:
```diff
@@ -69,13 +69,29 @@ static unsigned int get_page(off_t offset)
 {
+	int rc;
+#if defined(CONFIG_FLASH_STM32_ASYNC)
+	FLASH_STM32_PRIV(dev)->async_complete = false;
+	FLASH_STM32_PRIV(dev)->async_error = false;
+	HAL_FLASH_Program_IT(FLASH_TYPEPROGRAM_DOUBLEWORD, offset + FLASH_STM32_BASE_ADDRESS, val);
+	k_sem_take(&FLASH_STM32_PRIV(dev)->async_sem, K_FOREVER);
+	if (FLASH_STM32_PRIV(dev)->async_complete) {
+		LOG_DBG("Flash write successful. Wrote 0x%llx at 0x%lx",
+			val, offset + FLASH_STM32_BASE_ADDRESS);
+		rc = 0;
+	} else {
+		if (FLASH_STM32_PRIV(dev)->async_error) {
+			LOG_ERR("Flash write failed %d", FLASH_STM32_PRIV(dev)->async_ret);
+		}
+		rc = -EIO;
+	}
```
I wasn't 100% sure what you were looking for but I added line breaks before `#if`, and after `#else` and `#endif`. Let me know if that is not what you were looking for or if you wanted line breaks before and after `#else` or something else.

jpowen898 commented:
I just did some testing with this on the stm32h743 and it is failing to erase the correct sectors because of this issue https://github.com/zephyrproject-rtos/zephyr/issues/93714 where the flash size is being calculated incorrectly (4MB when it should be 2MB) and caused the `get_sector()` to return the wrong bank and sector numbers. It looks like there is a PR open to fix it already https://github.com/zephyrproject-rtos/zephyr/pull/93739.

jpowen898 commented on drivers/flash/flash_stm32f4x.c:
```diff
@@ -130,15 +162,42 @@ static int write_value(const struct device *dev, off_t offset, flash_prg_t val)
+		rc = 0;
+	} else {
+		if (FLASH_STM32_PRIV(dev)->async_error) {
+			LOG_ERR("Flash erase failed %d", FLASH_STM32_PRIV(dev)->async_ret);
```
Yes good call, that return code is actually a flash address.

jpowen898 commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -570,11 +596,34 @@ static int wait_write_queue(const struct flash_stm32_sector_t *sector)
+			if (FLASH_STM32_PRIV(dev)->async_error) {
+				LOG_ERR("Flash write failed %d", FLASH_STM32_PRIV(dev)->async_ret);
+			}
+			rc = -EIO;
```
Good idea. I will change that.

etienne-lms commented on drivers/flash/flash_stm32.c:
```diff
@@ -161,8 +161,16 @@ static int flash_stm32_read(const struct device *dev, off_t offset,
 
+#if defined(CONFIG_FLASH_STM32_ASYNC)
+	flash_stm32_sem_take(dev);
+#endif /* CONFIG_FLASH_STM32_ASYNC */
```
Prefer:
```c
	if (IS_ENABLED(CONFIG_FLASH_STM32_ASYNC)) {
		flash_stm32_sem_take(dev);
	}

etienne-lms commented on drivers/flash/flash_stm32.c:
```diff
@@ -458,6 +492,13 @@ static int stm32_flash_init(const struct device *dev)
+#if defined(CONFIG_FLASH_STM32_ASYNC)
+	flash_dev = FLASH_STM32_PRIV(dev);
+	flash_stm32_async_sem_init(dev);
+	IRQ_CONNECT(FLASH_IRQn, 0, flash_stm32_irq_handler, NULL, 0);
```
For portability (multi device), it may be worth to pass the device instance as argument to `flash_stm32_irq_handler()` instead of using unique global `flash_dev` reference.

jpowen898 replied:
I thought about doing this but the stm32 hal makes that tricky. Their hal is setup such that you have a single flash IRQ handler function `HAL_FLASH_IRQHandler()` which then can call the 2 end of operation callback functions (`HAL_FLASH_EndOfOperationCallback()` and `HAL_FLASH_OperationErrorCallback`). The problem is that those 2 callback functions are declared in the stm32 hal and then the user of the hal defines those functions and it figures it out at link time. So I can't change those function declarations to pass in the `flash_dev` as an arg without editing the stm32 hal directly.

etienne-lms commented on drivers/flash/flash_stm32f4x.c:
```diff
@@ -36,9 +36,21 @@ typedef uint8_t flash_prg_t;
+#endif
+#endif /* CONFIG_FLASH_STM32_ASYNC */
+
+bool flash_stm32_valid_range(const struct device *dev, off_t offset, uint32_t len, bool write)
```
Nitpicking: discard this change on function declaration? (unrelated)

etienne-lms replied:
If you really want to, then prefer add a dedicated commit, to not mix new implementation changes and code beautification changes in the same commit.

jpowen898 replied:
Sounds good. I'll remove all the changes made by `clang-format` I was using it because I wasn't sure exactly what the style guide required and some of the github actions were failing on the coding guidelines check.

etienne-lms commented on drivers/flash/flash_stm32f4x.c:
```diff
@@ -130,15 +161,43 @@ static int write_value(const struct device *dev, off_t offset, flash_prg_t val)
+			LOG_ERR("Flash erase failed at 0x%x", FLASH_STM32_PRIV(dev)->async_ret);
+		}
+		rc = -EIO;
+	}
```
A explicit `return rc` (or even explcit `return 0;` and `return -EIO;`) would be be simpler, for readibilty, IMHO.
Not a strong opinion, feel free to discard this comment.

Ditto in `write_value()`.

jpowen898 replied:
I made this change in all the functions except for stm32 h7 flash_write() function because it uses a for loop to do several consecutive writes and so it made sense to keep the return code var

jpowen898 commented on drivers/flash/flash_stm32f4x.c:
```diff
@@ -180,13 +239,12 @@ static int erase_sector(const struct device *dev, uint32_t sector)
-int flash_stm32_block_erase_loop(const struct device *dev,
-				 unsigned int offset,
-				 unsigned int len)
+int flash_stm32_block_erase_loop(const struct device *dev, unsigned int offset, unsigned int len)
```
No, I can remove it if you'd like. I was just using the clang-format on all these files with the zephyr format config `zephyr/.clang-format` and it removed the indentations.

jpowen898 commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -61,7 +61,7 @@ LOG_MODULE_REGISTER(LOG_DOMAIN);
 #define NUMBER_OF_BANKS             2
 #else
-#define NUMBER_OF_BANKS             1
+#define NUMBER_OF_BANKS 1
```
This change was made by `clang-format` using the `zephyr/.clang-format` config, but I can revert if you would like.

jpowen898 commented on drivers/flash/flash_stm32f4x.c:
```diff
@@ -379,20 +433,16 @@ static const struct flash_pages_layout stm32f4_flash_layout[] = {
 #error "Unknown flash layout"
 #endif /* FLASH_SECTOR_TOTAL == 5 */
-#endif/* !defined(FLASH_SECTOR_TOTAL) */
+#endif /* !defined(FLASH_SECTOR_TOTAL) */
```
I kept this beautification change and reverted the rest

etienne-lms replied:
Actaully unrelated to the async. Could you make a dedicated commit for the coding style fixes you want to propose?

etienne-lms commented on drivers/flash/flash_stm32.c:
```diff
@@ -37,8 +37,8 @@ LOG_MODULE_REGISTER(flash_stm32, CONFIG_FLASH_LOG_LEVEL);
 	.write_block_size = FLASH_STM32_WRITE_BLOCK_SIZE,
-	/* Some SoCs (L0/L1) use an EEPROM under the hood. Distinguish
-	 * between them based on the presence of the PECR register. */
+/* Some SoCs (L0/L1) use an EEPROM under the hood. Distinguish
+ * between them based on the presence of the PECR register. */
```
Please discard this change. Indentation was fine (even if coding style was not :)

etienne-lms commented on drivers/flash/flash_stm32.c:
```diff
@@ -68,7 +68,7 @@ static int flash_stm32_check_status(const struct device *dev)
 		LOG_DBG("Status: 0x%08lx",
 			(unsigned long)FLASH_STM32_REGS(dev)->FLASH_STM32_SR &
-							FLASH_STM32_SR_ERRORS);
+						   FLASH_STM32_SR_ERRORS);
```
This change is unrelated and not very consistent in term of indenation.
Could you discard it?

jpowen898 replied:
Sorry, I didn't mean to commit this. I'll remove it.

etienne-lms commented on drivers/flash/flash_stm32.c:
```diff
@@ -391,14 +399,14 @@ static int flash_stm32_get_size(const struct device *dev, uint64_t *size)
+/* Getting clocks information from device tree description depending
+ * on the presence of 'clocks' property.
+ */
 #if DT_INST_NODE_HAS_PROP(0, clocks)
 	.pclken = {
-		.enr = DT_INST_CLOCKS_CELL(0, bits),
-		.bus = DT_INST_CLOCKS_CELL(0, bus),
-	}
+							     .enr = DT_INST_CLOCKS_CELL(0, bits),
+							     .bus = DT_INST_CLOCKS_CELL(0, bus),
+						     }
```
Please discard these 2 changes.

etienne-lms commented on drivers/flash/flash_stm32.c:
```diff
@@ -416,6 +424,35 @@ static DEVICE_API(flash, flash_stm32_api) = {
+	}
+}
+
+/* cube hal functions for end of a async flash operation */
```
```suggestion
/* STM32 HAL functions for end of a async flash operation */

etienne-lms commented on drivers/flash/flash_stm32f4x.c:
```diff
@@ -36,9 +36,23 @@ typedef uint8_t flash_prg_t;
 bool flash_stm32_valid_range(const struct device *dev, off_t offset,
-			     uint32_t len,
-			     bool write)
+	uint32_t len,
+	bool write)
```
plaese discard these 2 unrelated changes.

etienne-lms reviewed (commented):
Check also the few comments posted by CI compliance checks bot.

etienne-lms reviewed (commented):
Minor issues, otherwise LGTM.

jpowen898 commented on drivers/flash/Kconfig.stm32:
```diff
@@ -102,4 +105,13 @@ config USE_MICROCHIP_QSPI_FLASH_WITH_STM32
+	select USE_STM32_HAL_FLASH_EX
+	help
+	  Use asynchronous flash operations to unblock other threads while
+	  flash is busy.
```
So yes when you enable this all operations are done in "asynchronous" mode, I am happy to document that if you would like, just point me to the correct spot to do that. But to be clear the `flash_erase()` and `flash_write()` functions complete the same high level tasks as they did before and return in the same amount of time so it should be pretty transperent to the user of the function if it is in async mode or not. Also those functions were never ISR safe because `flash_erase()` takes ~1.1s to return (longer if you erase multiple sectors at a time) and `flash_write()` takes ~3.5ms to return.

jpowen898 committed:
drivers: Implement async flash transactions for stm32's

Added optional use of the stm32 hal async flash write and erase
capabilities on the l4, f4, and h7 soc families. It is implemented
using semaphores to allow for other threads to take action during
the erase/write actions. This feature is disabled by default and
can be optionally enabled using the FLASH_STM32_ASYNC config option.

Signed-off-by: Parker Owen <jpowen898@gmail.com>

etienne-lms reviewed (commented):
Commit message says 
> This feature is disabled by default and
> can be optionally enabled using the FLASH_STM32_ASYNC config option

Actually the feature is default enabled for drivers that support async operations.
**(edited) my mistake, see my comment below.**

Worth to mention also that when enabled, read operations are synced on flash erase/write semaphore.

~~When so, read operations are se~~

etienne-lms commented:
> Commit message says
> > This feature is disabled by default and
> > can be optionally enabled using the FLASH_STM32_ASYNC config option
>
> Actually the feature is default enabled for drivers that support async operations.

My mistake, it is default disabled. Discard this comment. Sorry.
